File: android/feature-ui/src/main/java/com/example/hearablemusicplayer/ui/library/viewmodel/SongDetailViewModel.kt
package com.example.hearablemusicplayer.ui.library.viewmodel

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope

import com.example.hearablemusicplayer.domain.music.MusicInfo
import com.example.hearablemusicplayer.domain.music.MusicLabel
import com.example.hearablemusicplayer.domain.music.usecase.GetDailyMusicRecommendationUseCase
import com.example.hearablemusicplayer.domain.setting.model.DailyMusicInfo
import com.example.hearablemusicplayer.domain.setting.model.PlaybackHistory
import com.example.hearablemusicplayer.domain.setting.usecase.PlaybackHistoryUseCase
import com.example.hearablemusicplayer.ui.common.util.UiState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SongDetailData(
    val musicInfo: MusicInfo,
    val dailyMusicInfo: DailyMusicInfo?,
    val labels: List<MusicLabel?>,
    val playbackHistory: List<PlaybackHistory> = emptyList()
)

@HiltViewModel
class SongDetailViewModel @Inject constructor(
    private val getDailyRecommendationUseCase: GetDailyMusicRecommendationUseCase,
    private val playbackHistoryUseCase: PlaybackHistoryUseCase,
    private val savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val _uiState = MutableStateFlow<UiState<SongDetailData>>(UiState.Idle)
    val uiState: StateFlow<UiState<SongDetailData>> = _uiState.asStateFlow()

    private var currentMusicId: Long? = null

    init {
        loadSongDetailFromRoute()
    }

    private fun loadSongDetailFromRoute() {
        try {
            // 尝试从 savedStateHandle 获取 musicId
            val musicId = savedStateHandle.get<Long>("musicId")
            if (musicId != null) {
                currentMusicId = musicId
                loadSongDetail(musicId)
            } else {
                _uiState.value = UiState.Error("无法获取歌曲 ID")
            }
        } catch (e: Exception) {
            _uiState.value = UiState.Error("导航参数错误: ${e.message}")
        }
    }

    fun loadSongDetail(musicId: Long) {
        currentMusicId = musicId
        _uiState.value = UiState.Loading
        
        viewModelScope.launch {
            try {
                val recommendation = getDailyRecommendationUseCase.getMusicWithExtraById(musicId)
                if (recommendation?.musicInfo != null) {
                    val historyFlow = playbackHistoryUseCase.getPlaybackHistory(musicId, 5)
                    
                    historyFlow.collectLatest { history ->
                        _uiState.value = UiState.Success(
                            SongDetailData(
                                musicInfo = recommendation.musicInfo!!,
                                dailyMusicInfo = recommendation.dailyMusicInfo,
                                labels = recommendation.labels,
                                playbackHistory = history
                            )
                        )
                    }
                } else {
                    _uiState.value = UiState.Error("Music not found")
                }
            } catch (e: Exception) {
                _uiState.value = UiState.Error(e.message ?: "Unknown error")
            }
        }
    }

    fun retry() {
        currentMusicId?.let { loadSongDetail(it) }
    }
}

File: ios/HMP/HMP/Features/ViewModel/SongDetailViewModel.swift
import Foundation
import shared

@Observable
class SongDetailViewModel {
    var state: UiState<SongDetailData_> = .idle
    private var currentMusicId: Int64?

    private let recommendationUseCase: GetDailyMusicRecommendationUseCase
    private let playbackHistoryUseCase: PlaybackHistoryUseCase

    init() {
        self.recommendationUseCase = KoinHelperKt.getGetDailyMusicRecommendationUseCase()
        self.playbackHistoryUseCase = KoinHelperKt.getPlaybackHistoryUseCase()
    }

    func load(musicId: Int64) {
        currentMusicId = musicId
        state = .loading
        Task {
            do {
                let rec = try await recommendationUseCase.getMusicWithExtraById(musicId: musicId)
                if let musicInfo = rec?.musicInfo {
                    let history = try await KoinHelperKt.getPlaybackHistory(
                        musicId: musicId, limit: 5
                    )
                    await MainActor.run {
                        self.state = .success(SongDetailData_(
                            musicInfo: musicInfo,
                            dailyMusicInfo: rec?.dailyMusicInfo,
                            playbackHistory: history
                        ))
                    }
                } else {
                    await MainActor.run { self.state = .error("Music not found") }
                }
            } catch {
                await MainActor.run { self.state = .error(error.localizedDescription) }
            }
        }
    }

    func retry() {
        if let id = currentMusicId { load(musicId: id) }
    }
}

struct SongDetailData_ {
    let musicInfo: MusicInfo_
    let dailyMusicInfo: DailyMusicInfo_?
    let playbackHistory: [PlaybackHistory_]
}
